consensus/src/dag/dag_state_sync.rs

```diff
@@ -0,0 +1,185 @@
+    }
+
+    /// Fast forward in the decoupled-execution pipeline if the block exists there
+    pub async fn sync_to_highest_commit_cert(&self, ledger_info: &LedgerInfoWithSignatures) {
+        let dag_reader = self.dag_store.read();
+
+        // if the anchor exists between ledger info round and highest ordered round
+        // Note: ledger info round <= highest ordered round
+        if dag_reader
+            .highest_committed_anchor_round()
+            < ledger_info.commit_info().round()
+            && dag_reader
+                .highest_ordered_anchor_round()
+                .unwrap_or_default()
+                >= ledger_info.commit_info().round()
+        {
+            self.notifier.send_commit_proof(ledger_info.clone()).await
+        }
+    }
+
+    /// Check if we're far away from this ledger info and need to sync.
+    /// This ensures that the block referred by the ledger info is not in buffer manager.
+    pub fn need_sync_for_ledger_info(&self, li: &LedgerInfoWithSignatures) -> bool {
+        let dag_reader = self.dag_store.read();
+        // check whether if DAG order round is behind the given ledger info round
+        // (meaning consensus is behind) or
+        // the highest committed anchor round is 2*DAG_WINDOW behind the given ledger info round
+        // (meaning execution is behind the DAG window)
+        (dag_reader
+            .highest_ordered_anchor_round()
+            .unwrap_or_default()
+            < li.commit_info().round())
+            || dag_reader
+                .highest_committed_anchor_round()
+                + STATE_SYNC_WINDOW_MULTIPLIER * DAG_WINDOW
+                < li.commit_info().round()
+    }
+
+    pub async fn try_sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeMessage,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        // Check whether to actually sync
+        let commit_li = node.ledger_info();
+        if !self.need_sync_for_ledger_info(commit_li) {
+            return Ok(None);
+        }
+
+        let dag_fetcher = Arc::new(DagFetcher::new(
+            self.epoch_state.clone(),
+            self.network.clone(),
+            self.time_service.clone(),
+        ));
+
+        self.sync_to_highest_ordered_anchor(node, dag_fetcher).await
+    }
+
+    /// Note: Assumes that the sync checks have been done
+    pub async fn sync_to_highest_ordered_anchor(
+        &self,
+        node: &CertifiedNodeMessage,
+        dag_fetcher: Arc<impl TDagFetcher>,
+    ) -> anyhow::Result<Option<Arc<RwLock<Dag>>>> {
+        let commit_li = node.ledger_info();
+
+        if commit_li.ledger_info().ends_epoch() {
+            self.notifier
+                .send_epoch_change(EpochChangeProof::new(
+                    vec![commit_li.clone()],
+                    /* more = */ false,
+                ))
+                .await;
+            // TODO: make sure to terminate DAG and yield to epoch manager
```

Comment:
this will come the other direction, epoch manager will terminate the old DAG instance after entering the new epoch